File: sorting/counting-sort/counting-sort.py
def countingSort(input):
   
    maxKey= max(input, key=lambda item:item[0])[0]
    
    bookeepingLength = maxKey+1
    bookeeping = [0] * bookeepingLength

    # Count keys frequency
    for item in input: 
        bookeeping[item[0]] += 1

    # at the end each element is the index 
    # of the last element with that key
    for i in range(1, bookeepingLength):
        bookeeping[i] += bookeeping[i-1] 

    output = [0] * len(input)

    # build sorted output iterating backward
    i = len(input) - 1
    while i >= 0:
        item = input[i]
        bookeeping[item[0]] -= 1
        position = bookeeping[item[0]]
        output[position] = item
        i -= 1

    return output

def main():
    print("Counting Sort Test")
    input = [(1, "a"), (1, "b"), (2, "c"), (4, "d"), (3, "e"), (0, "f"), (1, "g"), (2, "h"), (3, "i"), (1, "l")]
    print(input)
    output = countingSort(input)
    print(output)

main()

File: sorting/counting-sort/counting-sort.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Algorithms
{
    class CountingSort {

        private static int[] PartialSum(int[] input)
        {
            for (int i = 1; i < input.Count(); i++)
            {
                input[i] = input[i] + input[i - 1];
            }

            return input;
        }

        public static (int, string)[] CountingSort((int, string)[] items)
        {
            int max_key = items.Max(t => t.Item1);
            var bookkeeping = new int[max_key + 1];

            //count keys frequency
            foreach (var item in items) {
                bookkeeping[item.Item1]++;
            }

            //at the end each element is the index of the last element with that key
            bookkeeping = PartialSum(bookkeeping);

            var output = new (int, string)[items.Length];

            //build sorted output iterating backward
            for (int i = items.Length - 1; i >= 0; i--) {
                output[--bookkeeping[items[i].Item1]] = items[i];
            }

            return output;
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Counting Sort Test");
 
            var input = new (int, string)[] { (1, "a"), (1, "b"), (2, "c"), (4, "d"), (3, "e"), (0, "f"), (1, "g"), (2, "h"), (3, "i"), (1, "l") };
            
            Console.WriteLine("Input");
            foreach (var item in output) {
                Console.WriteLine("(" << item.Item1 << ", " << item.Item2 << ")");
            }
            
            var output = CountingSort(input);

            Console.WriteLine("Output");
            foreach (var item in output) {
                Console.WriteLine("(" << item.Item1 << ", " << item.Item2 << ")");
            }
        }
    }
}